contrib/envconfig/client_config.go

```diff
@@ -0,0 +1,210 @@
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package envconfig contains utilities to load configuration from files and/or environment variables.
+//
+// WARNING: Environment configuration is currently experimental.
+package envconfig
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/converter"
+)
+
+// ClientConfig represents a client config file.
+//
+// WARNING: Environment configuration is currently experimental.
+type ClientConfig struct {
+	Profiles map[string]*ClientConfigProfile
+}
+
+// ClientConfigProfile is profile-level configuration for a client.
+//
+// WARNING: Environment configuration is currently experimental.
+type ClientConfigProfile struct {
+	Address   string
+	Namespace string
+	APIKey    string
+	TLS       *ClientConfigTLS
+	Codec     *ClientConfigCodec
+	// When loading from TOML and env var, or writing to TOML, the keys are lowercased and hyphens are replaced with
+	// underscores. This is used for deduplicating/overriding too, so manually set values that are not normalized may
+	// not get overridden with [ClientConfigProfile.ApplyEnvVars].
+	GRPCMeta map[string]string
+}
+
+// ClientConfigTLS is TLS configuration for a client.
+//
+// WARNING: Environment configuration is currently experimental.
+type ClientConfigTLS struct {
+	Disabled                bool
+	ClientCertPath          string
+	ClientCertData          []byte
+	ClientKeyPath           string
+	ClientKeyData           []byte
+	ServerCACertPath        string
+	ServerCACertData        []byte
+	ServerName              string
+	DisableHostVerification bool
+}
+
+// ClientConfigCodec is codec configuration for a client.
+//
+// WARNING: Environment configuration is currently experimental.
+type ClientConfigCodec struct {
```

Comment:
Maybe call this ClientConfigRemoteCodec? What, if anything, could ever be configured here for local codec? And, even then, maybe it'd make sense to have them be different sub-configs

Reply:
It is the `codec` field in config and `TEMPORAL_CODEC_` prefixed env vars. So I want it to match those. If we want though, I could change the config field name, env var prefix, struct field name, and this type name.